docs/architecture/psa-driver-interface.md

```diff
@@ -0,0 +1,561 @@
+
+For transparent drivers, the init function does not take any parameter.
+
+For opaque drivers, the init function takes the following parameters:
+
+* `uint8_t *persistent_state`: the driver's persistent state. On the first boot of the device, this contains all-bits-zero. On subsequent boots, the core loads the last saved state.
+* `size_t persistent_state_size`: the size of the persistent state in bytes.
+
+For an opaque driver, if the init function succeeds, the core saves the updated persistent state. If the init function fails, the persistent state is unchanged.
+
+### Combining multiple drivers
+
+To declare a cryproprocessor can handle both cleartext and plaintext keys, you need to provide two driver descriptions, one for a transparent driver and one for an opaque driver. You can use the mapping in capabilities' `"names"` property to arrange for driver functions to map to the same C function.
+
+## Transparent drivers
+
+### Key format for transparent drivers
+
+The format of a key for transparent drivers is the same as in applications. Refer to the documentation of `psa_export_key()` and `psa_export_public_key()`.
+
+### Key management with transparent drivers
+
+Transparent drivers may provide the following key management functions:
+
+* `"generate_key"`: called by `psa_generate_key()`, only when generating a key pair (key such that `PSA_KEY_TYPE_IS_ASYMMETRIC` is true).
+* `"derive_key"`: called by `psa_key_derivation_output_key()`, only when deriving a key pair (key such that `PSA_KEY_TYPE_IS_ASYMMETRIC` is true).
+* `"export_public_key"`: called by the core to obtain the public key of a key pair. The core may call this function at any time to obtain the public key, which can be for `psa_export_public_key()` but also at other times, including during a cryptographic operation that requires the public key such as a call to `psa_verify_message()` on a key pair object.
+
+Transparent drivers are not involved when importing, exporting, copying or destroying keys, or when generating or deriving symmetric keys.
+
+### Fallback
+
+If a transparent driver function is part of a capability which has a true `"fallback"` property and returns `PSA_ERROR_NOT_SUPPORTED`, the built-in software implementation will be called instead. Any other value (`PSA_SUCCESS` or a different error code) is returned to the application.
+
+If there are multiple available transparent drivers, the core tries them in turn until one is declared without a true `"fallback"` property or returns a status other than `PSA_ERROR_NOT_SUPPORTED`.
+
+If a transparent driver function is part of a capability where the `"fallback"` property is false or omitted, the core should not include any other code for this capability, whether built in or in another transparent driver.
```

Comment:
Is fallback really a property of the accelerator? I'm thinking it would be more natural to have a global fallback option per feature set, for which it is the library end-user's responsibility to turn it on or off. It doesn't make sense that one accelerator plugin can ship with the fallback property set, and will pull in a whole bunch of fallback code for everyone using the plugin. Simultaneously, it doesn't make sense to ship a plugin without the fallback property set, since then the user cannot add the fallback bits he/she needs.

Reply:
No fallback means “I can handle everything I declare”. Fallback means “I can't handle everything I declare”. The application has no way to know whether falling back is needed. So yes, it's a property of the driver.

Reply:
Are the declarations not granular enough such that a driver should be able to define strictly what it supports?

Reply:
I hope they mostly are, but I'm sure a few exceptions will come up. Maybe a secure element with a limit on input size. Or an RSA accelerator that only works with e=65537.